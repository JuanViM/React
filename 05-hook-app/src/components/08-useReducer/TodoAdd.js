import React from 'react'
import { useForm } from '../../hook/useForm';

export const TodoAdd = ({handleAddTodo}) => {

    
    
    //esto es el estado del formulario
   const [{description},handleInputChange,reset] = useForm({
    description:""
});

    const handleSubmit = (e) =>{
        e.preventDefault();

        if( description.trim().length <=1){
            return;
        }
        
        const nuevaTarea = {
            id: new Date().getTime(),
            desc: description,
            done:false
        } 

        
        //le mandamos a dispatch la accion de aÃ±adir tarea
        handleAddTodo(nuevaTarea);
        reset();
    } 

    return (
        <>
            <h4>Agregar TODO</h4>
            <hr />
            <form onSubmit={handleSubmit}>
                <input
                    type="text"
                    name='description'
                    className='form-control'
                    placeholder='Aprender....'
                    autoComplete='off'
                    //vamos a darle la descripcion que introducimos en el formulario al input mediante el value
                    value={description}
                    onChange={handleInputChange}
                />
                <button
                    type="submit"
                    className='btn btn-outline-primary mt-1 btn-block'>
                    Agregar
                </button>

            </form>
        </>
    )
}
