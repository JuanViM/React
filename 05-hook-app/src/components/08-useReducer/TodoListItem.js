import React from 'react'

export const TodoListItem = ({todo, index, handleToggle, handleDelete}) => {
    return (
        <>
            <li
                key={todo.id}
                className="list-group-item">
                {/* Esto quiere decir que si todo.done es true entonces && retorna la clase complete 
                  sino no haga nada */}
                <p className={`${todo.done ? 'complete' : null}`}
                    onClick={() => handleToggle(todo.id)}> {index + 1}.  {todo.desc}</p>
                <button
                    className='btn btn-danger'
                    onClick={() => handleDelete(todo.id)}>
                    borrar

                </button>
            </li>
            </>
    )
}
