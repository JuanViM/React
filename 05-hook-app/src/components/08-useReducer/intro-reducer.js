
//aqui tendriamos las tareas que tenemos por hacer
const tareas = [{
    id:1,
    todo:"comprar pan",
    done:false,
}];
//a esta funcion le vamos a mandar el state o tarea y una funcion llamda accion que sera lo que modificara el state
//que por sera igual a las tareas
const todoReducer = (state = tareas,action) => {
 //con la ? le decimos que si tiene algun valor lea el tipo y que sino no haga nada
    if (action?.type == "agregar" ) {
        //para agregarlo usaremos el operador spreat para esparcir todos los elementos que tenemos en el array "...state" y luego añadirle los nuevos "action.payload" 
        return [...state,action.payload]
    }
    return state;
};

let listaTareas = todoReducer();

//para añadir una nueva tarea lo haremos d ela siguiente manera

const tarea2 = {
    id:2,
    todo:"comprar leche",
    done:false,
}

// las acciones son objetos,debemos darle el nombre de payload como recomendacion y el type seria lo que va a hacer, en este caso agregar
const AgregarTareas = {
    type: "agregar",
    payload:tarea2,
}

listaTareas = todoReducer( tareas, AgregarTareas);

console.log(listaTareas);