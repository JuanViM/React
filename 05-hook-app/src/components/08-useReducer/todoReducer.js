
export const todoReducer = (state = [],action) => {

    switch (action.type) {
        case "add":
            //volvemos a usar el operador spread para rellenar el array con lo que tenia y le aÃ±adimos lo nuevo
            return [...state,action.payload];
            //vamos a borrar un todo, el filter va devolver todos los elementos de state que cumplan una condicion
            // la condicion seria que todo.id sea diferente al todo que yo quiera borrar
        case "delete":
            return state.filter( todo =>todo.id != action.payload);


        case "realizado":
            //vamos a utilizar un map para recorrer cada uno de los elementos y retornar un valor
            //comentamos esta linea para simplificar el codigo
            // return state.map( todo => {
          
            //     //si el todo.id introducido es igual al action.payload
            //     if(todo.id == action.payload){
            //         return {
            //             ...todo,
            //             done:!todo.done,
            //         }
            //         }else {
            //             return todo;
            //         }
            // });
            //lo simplificamos de la siguiente manera
            //con esto mapeamos igual que el anterior pero buscamos de la siguiente manera
            //si el todo introducido el id coincide con un id de dentro del state 
            //entonces retorname el array de todo y el todo con done en false
            //sino retorname el array,todo esto realizado mediante un operador ternario
            return state.map( todo =>
                (todo.id ==action.payload)?{...todo,done:!todo.done} : todo)
        default:
            return state;
    }
}