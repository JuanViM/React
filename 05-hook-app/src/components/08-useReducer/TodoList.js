import React from 'react';
import { TodoListItem } from './TodoListItem';



export const TodoList = ({ todo, handleToggle, handleDelete }) => {
    return (
        <>
            <ul className='list-group list-group-flush'>


                {
                    //vamos a mapear los todo para retornar toda la lista de objetos que esta tenga ademas debemos de ponerle el key que seria el id
                    todo.map((todo, i) => (
                        <TodoListItem
                            key={todo.id} 
                            todo={todo}
                            index={i}
                            handleToggle={handleToggle}
                            handleDelete={handleDelete} />
                    ))
                }
            </ul>
        </>
    )
}
