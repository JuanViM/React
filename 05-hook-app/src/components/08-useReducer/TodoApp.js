import React, { useEffect, useReducer } from 'react';
import { useForm } from '../../hook/useForm';
import "./styles.css";
import { TodoAdd } from './TodoAdd';
import { TodoList } from './TodoList';
import { todoReducer } from './todoReducer';


//la funcion init va a devolver el objeto que estaba en initialState, ahora no tenemos el initialState asi que 
//en  const [todo,dispatch] debemos mandarlo como un arreglo vacio. "const [todo,dispatch] = useReducer(todoReducer, [],init);"
const init = ()=>{

//vamos a retornar el localStorage de la siguiente manera, si no es null retorna eso, si es null retorna un array vacio
    return JSON.parse(localStorage.getItem("todo")) || [];

    // return [{
    //     id:new Date().getTime(),
    //     desc:"aprender React",
    //     done:false,
    // }];

}

export const TodoApp = () => {

    //el dispatch es una funcion la cual le manda una accion y va a saber a que reducer hay que mandarle la informacion y va a redibujarlo
   //le mandamos al final una funcion llamada init que evitara que se ejecute siempre que hay cambios
    const [todo,dispatch] = useReducer(todoReducer,[],init);


    //vamos a realizar un useEffect para que cada vezx que los todos cambien se guarden en el localStorage
    useEffect(() => {
      //vamos a guardar en el localStorage pero este no guardo objetos solo string por lo tanto debemos pasarlo a JSON.stringlify
      localStorage.setItem("todo",JSON.stringify(todo));
    }, [todo]);

    const handleDelete = (todoId)=>{

        const borrarTarea = {
            type:"delete",
            payload:todoId,
        }
        dispatch(borrarTarea);
       
    }
    
    const handleToggle = (todoId) => {
        
        //vamos a hacerlo sinc rear const mandandole el objeto del tiron
        dispatch({
            type:"realizado",
            payload:todoId
        });
    }

    const handleAddTodo = (newTodo) =>{
         
        //le mandamos a dispatch la accion de a√±adir tarea
        dispatch({
            type:"add",
            payload:newTodo,
        });
    }
    
    
   
  return (
    <div>
        <h1>TodoApp <small>({todo.length})</small></h1>
        <hr />

    <div className='row'>
        <div className='col-7'>
        <TodoList 
        todo={todo}
        handleDelete={handleDelete}
        handleToggle={handleToggle}        
        />
        </div>

        <div className='col-5'>
            
        Todo
        <TodoAdd 
        handleAddTodo={handleAddTodo}/>
        </div>

    </div>

    </div>
  )
}
