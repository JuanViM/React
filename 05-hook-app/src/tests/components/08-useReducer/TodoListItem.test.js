import { shallow } from "enzyme/build";
import { TodoListItem } from "../../../components/08-useReducer/TodoListItem";
import { demoTodos } from "../../fixtures/demoTodos";

describe('pruebas en <TodoListItem/>', () => { 

    const handleDelete = jest.fn();
    const handleToggle = jest.fn();
    const wrapper = shallow(<TodoListItem 
    todo={demoTodos[0]}
    index={0}
    handleToggle={ handleToggle }
    handleDelete={ handleDelete }
    />);

    test('debe mostrarse correctamente', () => { 
        
        expect(wrapper).toMatchSnapshot();

     })

     test('Debe de llamar a la funcion handleDelete', () => { 
         
        wrapper.find("button").simulate("click");
        expect(handleDelete).toHaveBeenCalledWith(demoTodos[0].id);

      })

      test('debe de llamar a la funcion handleToggle', () => { 
          
        wrapper.find("ptodo").simulate("click");
        expect(handleToggle).toHaveBeenCalledWith(demoTodos[0].id);

       });

       test('debe de mostrar el texto correctamente', () => { 
           
        const p = wrapper.find("p");
        expect (p.text().trim()).toBe(`1. ${demoTodos[0].desc}`)

        });
 })