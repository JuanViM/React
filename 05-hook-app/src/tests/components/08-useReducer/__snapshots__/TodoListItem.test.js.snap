// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en <TodoListItem/> debe mostrarse correctamente 1`] = `
<Fragment>
  <li
    className="list-group-item"
    key="1"
  >
    <p
      className="null"
      onClick={[Function]}
    >
       
      1
      .  
      Aprender React
    </p>
    <button
      className="btn btn-danger"
      onClick={[Function]}
    >
      borrar
    </button>
  </li>
</Fragment>
`;
